README.md
# SMSGateway – A flat-file based SMS gateway PHP class using an open source Android app

## Features
- SMS gateway using the open source Android app [SMS Gateway](https://github.com/medic/cht-gateway/releases/latest) - use the generic release

## License
This software is distributed under the [LGPL-3.0-only](https://www.gnu.org/licenses/lgpl-3.0.html) license. Please read [LICENSE](https://github.com/multiOTP/SMSGateway/blob/main/LICENSE) for information on the software availability and distribution.

## Installation & loading
SMSGateway is available on [Packagist](https://packagist.org/packages/multiOTP/SMSGateway) (using semantic versioning), and installation via [Composer](https://getcomposer.org) is the recommended way to install SMSGateway. Just add this line to your `composer.json` file:

```json
"multiotp/smsgateway": "^1.0"
```

or run

```sh
composer require multiotp/smsgateway
```

Note that the `vendor` folder and the `vendor/autoload.php` script are generated by Composer; they are not part of SMSGateway.

Alternatively, if you're not using Composer, you
can [download SMSGateway as a zip file](https://github.com/multiOTP/SMSGateway/archive/master.zip), then copy the contents of the SMSGateway folder into one of the `include_path` directories specified in your PHP configuration and load each class file manually:

```php
<?php
use multiOTP\SMSGateway\SMSGateway;

require 'path/to/SMSGateway/src/SMSGateway.php';
```

## Two simple examples

```php
<?php
//Import SMSGateway classes into the global namespace
//These must be at the top of your script, not inside a function
use multiOTP\SMSGateway\SMSGateway;

//Load Composer's autoloader
require 'vendor/autoload.php';

//Create an instance
$smsgateway = new SMSGateway();

// Set the data folder
$smsgateway->setDataPath(__DIR__ . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR);

// Set the shared secret
$smsgateway->setSharedSecret("secret");

$device_id   = "demo";
$to = "+1234567890";
$message = "Demo message";
$device_h = $smsgateway->calculateAuthenticationHash($device_id);

$message_id = $smsgateway->sendMessage($device_id, $to, $message);

echo "Full URL for Android app URL: https://......./?id=$device_id&h=$device_h";
```

```php
<?php
//Import SMSGateway classes into the global namespace
//These must be at the top of your script, not inside a function
use multiOTP\SMSGateway\SMSGateway;

//Load Composer's autoloader
require 'vendor/autoload.php';

// Please note that implementing a new_message_handling callback function
//  will flag the new messages as read when returning from the function.
function new_message_handling($array) {
  // Handling $array of new received messages
  // [["device_id"    => "device id",
  //   "message_id"   => "message id",
  //   "from"         => "from phone number",
  //   "sms_sent"     => "sms_sent timestamp (ms)",
  //   "sms_received" => "sms_received timestamp (ms)",
  //   "content"      => "message content",
  //   "last_update"  => "last update timestamp (ms)",
  //   "status"       => "UNREAD|READ"
  //  ],
  //  [...]
  // ]
}

function update_handling($array) {
  // Handling $array of status updates for sent messages
  // [["device_id"    => "device id",
  //   "message_id"   => "message id",
  //   "to"           => "to phone number",
  //   "content"      => "content",
  //   "last_update"  => "last update timestamp (ms)",
  //   "status"       => "NEW|PUSHED|PENDING|SENT|DELIVERED|FAILED"
  //  ],
  //  [...]
  // ]
}

function timeout_handling($array) {
  // Handling $array of devices not seen during the last "DeviceTimeout" seconds
  // [["device_id"   => "device_id",
  //   "last_update" => "device_last_update"
  //  ],
  //  [...]
  // ]
}

//Create an instance
$smsgateway = new SMSGateway();

// Set the data folder
$smsgateway->setDataPath(__DIR__ . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR);

// Launch the API server with callback functions definition
$smsgateway->apiServer("secret", "new_message_handling", "update_handling", "timeout_handling");
```

## SMS messages status

### Received message
- UNREAD: message has never been read on the gateway
- READ: message has already been read on the gateway

### Sent message
- NEW: new created message, not transfered yet to the Android device
- PUSHED: message pushed to the Android device
- PENDING: message is pending in the Android device
- SENT: message has been sent to the gateway's network
- DELIVERED: message has been received by the recipient's phone
- FAILED: message delivery has failed and will not be retried
- MISSING: message with this message id is missing, no state available

## Online demo
A full working gateway implementation is available here : [Online SMSGateway demo](https://1-2-3-4-5-6.net/smsgateway/). Click the link and everything is self-explanatory. You will simply have to install and configure the companion open source Android app in order to send and receive SMS messages through this demo gateway (as explained after sending a first SMS message using the online demo gateway).  
When sending a message, the following information will be returned in the http header and in the html meta tags:
- X-SMSGateway-State: state of the sent message (NEW|FAILED)
- X-SMSGateway-State-Url: full url to check the state of the message
- X-SMSGateway-Message-Id: message id

## Documentation
Example of how to use SMSGateway for a common scenario can be found in the [examples](https://github.com/multiOTP/SMSGateway/tree/main/examples) folder. If you're looking for a good starting point, we recommend you start with [the gateway example](https://github.com/multiOTP/SMSGateway/tree/main/examples/gateway.php).

## Android specific parameters
Using adb shell, you should change these two parameters:
- sms_outgoing_check_max_count
- sms_outgoing_check_interval_ms

Example for a limit of 200 messages per minute:

```
adb shell
settings put global sms_outgoing_check_max_count 200
settings put global sms_outgoing_check_interval_ms 60000
```

You need to reboot your Android phone. after these changes

## Changelog
See [CHANGELOG](CHANGELOG.md).

That's it. You should now be ready to use SMSGateway!

CHANGELOG.md
# SMSGateway Change Log

## Version 1.0.9 (2023-09-20)

### Fixed bugs
* PHP 8.1 fix (thanks to ratmole)

## Version 1.0.8 (2023-03-30)

### Fixed bugs
* Better json decoding, ignoring chr(13)

## Version 1.0.6 (2023-03-21)

### Enhancements
* Enhanced gateway example

## Version 1.0.5 (2022-10-17)

### Enhancements
* Enhanced gateway example

## Version 1.0.4 (2022-10-13)

### Enhancements
* More sent messages attributes
* Full gateway example

### Fixed bugs
* Descending sorted messages

## Version 1.0.3 (2022-10-11)
* First public version

## Version 0.0.1 (2022-09-10) (unreleased)
* Initial implementation


examples/gateway.php
<?php
/**
 * @file  gateway.php
 * @brief Simple gateway, using the default shared secret
 *         This simple demo gateway is available here : https://1-2-3-4-5-6.net/smsgateway/
 *
 * PHP 5.3.0 or higher is supported.
 *
 * @author    Andre Liechti (SysCo systemes de communication sa) <info@multiotp.net>
 * @version   1.1.1
 * @date      2023-03-30
 * @since     2022-09-10
 * @copyright (c) 2022-2023 SysCo systemes de communication sa
 * @copyright GNU Lesser General Public License
 *
 *********************************************************************/

// Import SMSGateway classes into the global namespace
// These must be at the top of your script, not inside a function
use multiOTP\SMSGateway\SMSGateway;

// Some tricks to load the SMSGateway class in different situations
if (!class_exists('multiOTP\SMSGateway\SMSGateway')) {
  if (file_exists('../src/SMSGateway.php')) {
    // Quick load of SMSGateway without using composer
    require_once '../src/SMSGateway.php';
  } else {
    // Composer autoload
    require '../vendor/autoload.php';
  }
}

// Create an SMSGateway instance if not done yet, and define the flat-file data folder
if (!isset($smsgateway)) {
  $smsgateway = new SMSGateway();
  $smsgateway->setDataPath(__DIR__ . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR);
}

// Detect the URL
$url = (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? "https" : "http") . "://" .$_SERVER['HTTP_HOST'].($_SERVER['PHP_SELF']);

// Retrieve some parameters
$command = isset($_GET["m"]) ? "m" : (isset($_GET["i"]) ? "i" : (isset($_GET["e"]) ? "e" : ""));
$h = isset($_GET["h"]) ? $_GET["h"] : "";
$mid = isset($_GET["mid"]) ? $_GET["mid"] : "";

// Correct the international format of the phone number if needed
$to = isset($_GET["to"]) ? $_GET["to"] : "";
if ("00" == substr($to, 0, 2)) {
  $to = "+" . substr($to, 2, strlen($to) -2);
}

// Define a default message if needed
$message = isset($_GET["message"]) ? $_GET["message"] : ""; // Hello World 😉

// Retrieve the device id
$id = isset($_GET["id"]) ? $_GET["id"] : "";
$device_id = $id;
if ((!empty($to)) && empty($device_id)) {
  $device_id = substr(md5(uniqid("", true)), 0, 16);
} elseif ((empty($to)) && (!empty($device_id)) && (!file_exists($smsgateway->getDataPath() .$device_id))) {
  $device_id = "";
}

// Calculate the device hash based on the secret
$device_h = $smsgateway->calculateAuthenticationHash($device_id);

// Check if device hash is valid for an existing device, otherwise flush the device id
if ((!empty($id)) && ($h != $device_h)) {
  $device_id = "";
} else {
  $smsgateway->updateDataStructure($id);
}

if ((!empty($mid)) && (!empty($device_id))) {
  $message_state = "MISSING";
  $message_array = $smsgateway->readSentStatus($id, $mid);
  if (isset($message_array[0]['status'])) {
    $message_state = $message_array[0]['status'];
  }
  echo $message_state;

} elseif ("e" == $command) {

  // An enhanced command can be implemented here

} elseif (("m" == $command) && (!empty($device_id))) {

  // Display messages resume for the "m" command
  echo "<h1>SMSGateway ".$smsgateway->getVersion()."</h1>";
  echo "<h2>New SMS messages received</h2>";
  $new_messages_count = 0;
  foreach ($smsgateway->readNewMessages($id) as $message) {
    echo date("Y-m-d H:i:s", $message['sms_received'] / 1000)." ".$message['from'].": ".$message['content']. "<br />";
    $new_messages_count++;
  }
  if (0 == $new_messages_count) {
    echo "none<br />";
  }
  echo "<h2>All SMS messages received</h2>";
  $messages_count = 0;
  foreach ($smsgateway->readAllMessages($id) as $message) {
    echo date("Y-m-d H:i:s", $message['sms_received'] / 1000)." ".$message['from'].": ".$message['content']. "<br />";
    $messages_count++;
  }
  if (0 == $messages_count) {
    echo "none<br />";
  }
  echo "<h2>All SMS messages sent</h2>";
  $sent_messages_count = 0;
  foreach ($smsgateway->readAllSentStatus($id) as $message) {
    if ("DELIVERED" == $message['status']) {
      $pre_status = "<b>";
      $post_status = "</b>";
    } else {
      $pre_status = "";
      $post_status = "";
    }
    echo date("Y-m-d H:i:s", $message['last_update'] / 1000)." ".$message['to']." [" . $pre_status . "<a target=\"track_" . $message['message_id'] . "\" href=\"$url?id=$device_id&h=$h&mid=" . $message['message_id'] . "\">" . $message['status'] . "</a>" . $post_status . "]: ".$message['content']. "<br />";
    $sent_messages_count++;
  }
  if (0 == $sent_messages_count) {
    echo "none<br />";
  }

} elseif (empty($device_id) || ("i" == $command)) {

  // Display basic usage info
  if ("" == $to) {
    $autofocus_to = "autofocus=\"autofocus\"";
    $autofocus_message = "";
  } else {
    $autofocus_to = "";
    $autofocus_message = "autofocus=\"autofocus\"";
  }
  
  $output = "<html><head><title>SMSGateway ".$smsgateway->getVersion()."</title></head><body>\n";
  
  $output.= "<h1>SMSGateway ".$smsgateway->getVersion()."</h1>";
  $output.= "<br />";
  $output.= "<form action=\"".htmlspecialchars($_SERVER["PHP_SELF"])."\" method=\"get\">\n";
  if ("" != $id) {
    $output.= "  <label for=\"to\">Device identification: </label>\n";
    $output.= "  <input size=\"18\" type=\"text\" name=\"id\" placeholder=\"e.g. 01234567890abcdef\" value=\"$id\">\n";
    $output.= "<br />\n";
    if ("" != $h) {
      $output.= "  <label for=\"to\">Secret hash for the device: </label>\n";
      $output.= "  <input size=\"8\" type=\"text\" name=\"h\" placeholder=\"e.g. abcdef\" value=\"$h\">\n";
      $output.= "<br />\n";
    }
    $output.= "<br />";
  }
  $output.= "  <label for=\"to\">Destination mobile phone number: </label>\n";
  $output.= "  <input size=\"20\" $autofocus_to type=\"tel\" name=\"to\" placeholder=\"e.g. 00123456789012\" value=\"$to\">\n";
  $output.= "<br />\n";
  $output.= "<br />\n";
  $output.= "  <textarea $autofocus_message columns=\"40\" rows=\"5\" placeholder=\"e.g. Please call me back asap !\" name=\"message\" autocomplete=\"on\" maxlength=\"300\" cols=\"80\" wrap=\"soft\">$message</textarea>\n";
  $output.= "  <br /><br />\n";
  $output.= "  <input type=\"submit\" value=\"Send SMS message\">\n";
  $output.= "  <br /><br />\n";
  if ("" == $h) {
    $output.= "... or send a first message calling this URL: <b>$url?to=001234567890&message=Hello+world</b>";
  } else {
    $output.= "... or send a direct message calling this URL: <b>$url?id=$id&h=$h&to=".(("" != $to) ? $to : "001234567890")."&message=".urlencode(("" != $message) ? $message : "Hello world 🙂")."</b>";
  }
  $output.= "</form>\n";

  echo $output;

} elseif (!empty($to)) {
  
  // Push the message on the server
  $message_id = $smsgateway->sendMessage($device_id, $to, $message);
  
  $output = "<html><head><title>SMSGateway ".$smsgateway->getVersion()."</title>\n";

  if (empty($message_id)) {
    header('X-SMSGateway-State: FAILED');
    $output.= "<meta name=\"X-SMSGateway-State\" content=\"0\">\n";
  } else {
    header('X-SMSGateway-State: NEW');
    header('X-SMSGateway-State-Url: $url?id=$id&h=$h&mid=$message_id');
    header('X-SMSGateway-Message-Id: '.$message_id);
    $output.= "<meta name=\"X-SMSGateway-State\" content=\"NEW\">\n";
    $output.= "<meta name=\"X-SMSGateway-State-Url\" content=\"$url?id=$id&h=$h&mid=$message_id\">\n";
    $output.= "<meta name=\"X-SMSGateway-Message-Id\" content=\"$message_id\">\n";
  }

  $output.= "</head><body>\n";

  // Display usage information
  $output.= "<h1>SMSGateway ".$smsgateway->getVersion()."</h1>";
  $output.= "Message '<i>$message</i>' for '$to' with id '<b>$message_id</b>' pushed on the server.";
  $output.= "<br /><br />";
  $output.= "If not done yet, please install the Android SMSGatewayApp available here : <a href=\"https://github.com/medic/cht-gateway/releases/latest\" target=\"blank\">medic-gateway-X.X.X-generic-release.apk</a>";
  $output.= "<br /><br />";
  $output.= "URL to set in the Settings of the Android App: <b>$url?id=$device_id&h=$device_h</b>";
  $output.= "<br />";
  $output.= "<i>(don't forget to enable polling)</i>";
  $output.= "<br /><br />";
  $output.= "<a target=\"_sms_gateway_check\" href=\"$url?id=$device_id&h=$device_h&m\"><button>Click here to check SMS messages</button></a>";
  $output.= "<br /><br />";
  $output.= "<a target=\"_sms_send\" href=\"$url?id=$device_id&h=$device_h&to=&message=&i\"><button>Click here to send more SMS messages</button>";
  
  $output.= "\n";
  $output.= "</body>\n";
  $output.= "</html>\n";
  
  echo $output;
  
} else {

  // Run the API server
  $smsgateway->apiServer();
}